File: AlgorithmsPackPythonSolutions/Jump Game III.py
class Solution:
    def process(self, q, visited, arr, index):
        if index < 0 or index >= len(arr) or visited[index]:
            return False
        if arr[index] == 0:
            return True
        q.append(index)
        visited[index] = True
        return False
    
    def canReach(self, arr: List[int], start: int) -> bool:
        n = len(arr)
        if arr[start] == 0:
            return True
        
        q = collections.deque()
        q.append(start)
        visited = [False] * n
        visited[start] = True
        
        while q:
            cur = q.popleft()
            
            if self.process(q, visited, arr, cur + arr[cur]) or self.process(q, visited, arr, cur - arr[cur]):
                return True
        
        return False

File: AlgorithmsPackCPP/Jump Game III.cpp
class Solution {
public:
    bool process(queue<int>&q,vector<bool>&visited, vector<int>&arr, int index){
        if(index < 0 || index>=arr.size()||visited[index])
            return false;

        if(arr[index]==0)
            return true;

        q.push(index);
        visited[index] = true;
        return false;
    }

    bool canReach(vector<int>& arr, int start) {
        int n = arr.size();
        if(arr[start]==0)
            return true;

        queue<int>q;
        q.push(start);
        vector<bool>visited(n);
        visited[start] = true;

        for(int level = 0, sz = q.size(); !q.empty(); ++level, sz = q.size()){
            while(sz--){
                int cur = q.front();
                q.pop();

                if(process(q,visited,arr,cur+arr[cur])
                ||process(q,visited,arr,cur-arr[cur]))
                return true;
            }
        }
        return false;
    }
};
